e2e/registration_test.go

```diff
@@ -197,6 +200,54 @@ service_registration {
 	assert.False(t, registered)
 }
 
+// TestE2E_ServiceRegistration_InitError tests that if the initial run
+// of all tasks fails, then CTS is not registered.
+func TestE2E_ServiceRegistration_InitError(t *testing.T) {
+	setParallelism(t)
+	srv := newTestConsulServer(t)
+	defer srv.Stop()
+
+	tempDir := fmt.Sprintf("%s%s", tempDirPrefix, "registration_init_failure")
+	cleanup := testutils.MakeTempDir(t, tempDir)
+	defer cleanup()
+
+	// Use a nonexistent module
+	id := "cts-01"
+	port := testutils.FreePort(t)
+	configPath := filepath.Join(tempDir, configFile)
+	config := baseConfig(tempDir).appendID(id).appendPort(port).
+		appendConsulBlock(srv).appendTerraformBlock().
+		appendModuleTask("failing_task", "./test_modules/failing_module")
+	config.write(t, configPath)
+
+	cmd := exec.Command("consul-terraform-sync", fmt.Sprintf("--config-file=%s", configPath))
+	var buf bytes.Buffer
+	cmd.Stdout = &buf
+	cmd.Stderr = &buf
+	err := cmd.Start()
+	require.NoError(t, err)
+	defer cmd.Process.Signal(os.Interrupt)
+
+	// Wait until CTS has exited
+	exitCh := make(chan struct{})
+	go func() {
+		cmd.Wait()
+		exitCh <- struct{}{}
+	}()
+	select {
+	case <-exitCh:
+		break
+	case <-time.After(defaultWaitForTestReadiness):
+		t.Fatal("timed out waiting for CTS initialization to fail")
+	}
+
+	// Verify no attempt at registration and no service registered
+	output := buf.String()
+	assert.NotContains(t, output, "registering Consul-Terraform-Sync as a service with Consul")
```

Comment:
I was wondering how you would know it didn't try to register vs deregistered on exit. Nice!